Request for Comments: Added StringBuilder

cklein opened the pull request:
# RfC

This is a very early draft for an implementation of StringBuilder.

Initially I wanted StringBuilder to pre-allocate a list (with either a good default value or based on a parameter)
and use `PyList_SET_ITEM`.

```
typedef struct { 
    Py_ssize_t _lst; 
    Py_ssize_t _cap; 
    Py_ssize_t _idx; 
} HPyStringBuilder;
```

I changed it to use to match ListBuilder and TupleBuilder which are passed by value.
This forced me to use `PyList_Append`.
Checking the cpython implementation, it doesn't seem to be as bad as I initially thought, because
[list_resize](https://github.com/python/cpython/blob/master/Objects/listobject.c#L45) already has a
strategy for overallocation.

Just after finishing my branch, I started reading the implementation of `HPyTracker` which is similar to my initial idea.

On big difference between HPyTracker and this implementation for StringBuilder is that it maintains an array of HPy handles, while this implementation still uses a cpython list as a PyObject which stores other PyObject instances, similar to the other builders.

cklein committed:
Added StringBuilder

antocuni commented:
don't worry about opening PRs on the main hpy repo, you are very welcome to do so and it's easier for us to discuss design, ideas and implementations. For this specific case, it might be a good idea to start with an "API design" issue instead of a PR, because first we need to discuss how the API should look like.

Some random comments:
1. I think it should be called `UnicodeBuilder`, and we will also need `BytesBuilder` (no need to implement both at the same time, it's fine to do two separate PRs)

2. the most important thing to the right is the user facing API: the one you propose look reasonable, but I think it should be possible to pass a size to preallocate the buffer. E.g., currently users can call `PyUnicode_FromUnicode(NULL, 10)` and then fill the buffer manually: with HPy, this will be forbidden, so `UnicodeBuilder` needs to support this use case

3. the concrete implementation on CPython is less important at this stage, IMHO. Using a list of strings is good for now, but maybe in the future we will want to try different implementations (e.g. allocating a fixed-size buffer and fill it directly, using a `StringIO` object, etc. etc.)

4. as you say, you can use an approach similar to `HPyTracker` if you need to store multiple fields for a single builder.

Other random ideas for how the API should look like:
- if the user specifies a size and promises it cannot grow, we can probably use a faster implementation. Do we want to support this use case?
- maybe we could add `HPyUnicodeBuilder_Append_s(... const char* s)`
- what about `HPyUnicodeBuilder_SetSlice()`, to set only a specific portion of the buffer?

antocuni commented on hpy/devel/src/runtime/ctx_stringbuilder.c:
```diff
@@ -0,0 +1,63 @@
+    if (lst != NULL) {
+        PyObject *item = _h2py(h_item);
+        PyList_Append(lst, item);
+    }
```
you should check that `h_item` is a `unicode` and raise an error otherwise

Commit:
Renamed StringBuilder to UnicodeBuilder

cklein committed:
Implementation with a preallocated list

The idea turned out to be very closely related to HPyTracker,
so I blatantly stole some code there.

cklein commented:
OK, thanks a lot, I'm going to move discussions to the main project.
